### PR TITLE
Add batching code to FirebaseStorageManager to allow for infinite scroll

### PR DESCRIPTION
**Pull request by MikolJiron**
Add code for batching queries, change credentials to allow any batching queries to be made locally or deployed.

### CODE REVIEW

**Review by CarlosM10011 — APPROVED** [review]

**Review by Lorraine-Lyu — APPROVED** [review]

File: src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java [hunk 1]
```diff
@@ -160,4 +161,139 @@ public static String getCollection(Firestore database, String collectionToGetFro
...
+   * Either we send in the first requested batch, or we send the next batch after one has already been sent.
+   * @param database - The database to retrieve from.
+   * @param collectionToGetFrom - The name of the collection to retrieve the batch from.
+   * @param batchSizeLimit - The maximum size of the batch, which is the size of a new page of data in infinite scroll.
+   * @param lastDocID - The ID of the last object I sent from the most recent batch.
+   * @param parameterToSortBy - The parameter I want to presort my collection by.
+   * @return - The json string representing the new batch of documents to be sent to the frontend.
+   * @throws FirebaseException
+   */
+  public static String getCollectionBatch(Firestore database, String collectionToGetFrom, int batchSizeLimit,  String lastDocID, String parameterToSortBy) throws FirebaseException {
```
Comment (tinagrasso) [inline src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java]: line length

File: src/test/java/com/google/step/finscholar/firebase/FirebaseStorageManagerTest.java [hunk 2]
```diff
@@ -69,4 +70,25 @@ public void setMultipleDocumentsAndRetrieveCollection() {
     }
     Assert.assertNotEquals(ServletConstantValues.DEFAULT_VALUE, json);
   }
+
+  @Test
+  public void retrieveCollectionBatchWithIDandWithoutID() {
+    String jsonWithID = ServletConstantValues.DEFAULT_VALUE;
+    try {
+      jsonWithID = FirebaseStorageManager.getCollectionBatch(firebase, TEST_COLLECTION_NAME, TEST_BATCH_SIZE_LIMIT, TEST_DOCUMENT_NAME, ServletConstantValues.DEFAULT_VALUE);
```
Comment (tinagrasso) [inline src/test/java/com/google/step/finscholar/firebase/FirebaseStorageManagerTest.java]: line length.

File: src/test/java/com/google/step/finscholar/firebase/FirebaseStorageManagerTest.java [hunk 3]
```diff
@@ -69,4 +70,25 @@ public void setMultipleDocumentsAndRetrieveCollection() {
...
+    } catch (Exception e) {
+      System.out.println(e.toString());
+    }
+    System.out.println("Batch query results with ID: " + jsonWithID);
```
Comment (tinagrasso) [inline src/test/java/com/google/step/finscholar/firebase/FirebaseStorageManagerTest.java]: Use a real logger.

**Review by tinagrasso — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java [hunk 4]
```diff
@@ -171,4 +172,150 @@ public static String getCollection(Firestore database, String collectionToGetFro
...
+   * @param database - The database to retrieve from.
+   * @param collectionToGetFrom - The name of the collection to retrieve the batch from.
+   * @param batchSizeLimit - The maximum size of the batch, which is the size of a new page of 
+   *   data in infinite scroll.
+   * @param lastDocID - The ID of the last object I sent from the most recent batch.
+   * @param parameterToSortBy - The parameter I want to presort my collection by.
+   * @return - The json string representing the new batch of documents to be sent to the frontend.
+   * @throws FirebaseException
+   */
+  public static String getCollectionBatch(Firestore database, String collectionToGetFrom,
```
Comment (tinagrasso) [inline src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java:187]: No parameter validation?

File: src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java [hunk 5]
```diff
@@ -171,4 +172,150 @@ public static String getCollection(Firestore database, String collectionToGetFro
...
+      // Throws a FirebaseException if we can't connect to firestore.
+      throw new FirebaseException(ServletConstantValues.UNABLE_TO_READ_FROM_FIRESTORE, e);
+    }
+
+    // If the last document I sent still exists, then retrieve a batch of size=batchSizeLimit documents 
+    //   that occur after the last doc in the collection.
+    if (document.exists()) {
+      Query page = (parameterToSortBy.equals(ServletConstantValues.DEFAULT_VALUE) ? 
+          collectionReference.limit(batchSizeLimit) : 
+          collectionReference.orderBy(parameterToSortBy).limit(batchSizeLimit)).startAfter(document);
```
Comment (tinagrasso) [inline src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java]: Can you guarantee consistency from this query?

File: src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java [hunk 6]
```diff
@@ -171,4 +172,150 @@ public static String getCollection(Firestore database, String collectionToGetFro
...
+      return getCollectionQuery(page);
+
+    } else {
+      // Throws a Firebase Exception if the document does not exist.
+      String message = ServletConstantValues.DOCUMENT + lastDocID + ServletConstantValues.DNE;
```
Comment (tinagrasso) [inline src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java]: String formatter :)

```
"The document with Id %s does not exist.
```

**Review by tinagrasso — CHANGES_REQUESTED** [review]

**Review by hzimz — APPROVED** [review]

File: src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java [hunk 7]
```diff
@@ -169,4 +166,159 @@ public static String getCollection(Firestore database, String collectionToGetFro
...
+    // A batch size limit needs to be specified in order to make any query.
+    if (batchSizeLimit == 0) {
+      throw new FirebaseException(BATCH_SIZE_NOT_SPECIFIED);
+    }
+
+    CollectionReference collectionReference = database.collection(collectionToGetFrom);
+
+    // If the lastDocID is null, then we know this is the first batch to send.
+    // Else simply get the next batch in the collection.
+    if (lastDocID == null) {
```
Comment (tinagrasso) [inline src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java]: Use Optional to protect against NPEs.
https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html

File: src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java [hunk 8]
```diff
@@ -169,4 +166,159 @@ public static String getCollection(Firestore database, String collectionToGetFro
...
+    } catch (Exception e) {
+      // Throws a FirebaseException if we can't connect to firestore.
+      throw new FirebaseException(ServletConstantValues.UNABLE_TO_READ_FROM_FIRESTORE, e);
+    }
+
+    // If the last document I sent still exists, then retrieve a batch of size=batchSizeLimit documents 
+    //   that occur after the last doc in the collection.
+    // If parameterToSortBy is null, then don't sort the query.
+    if (document.exists()) {
+      Query page = (parameterToSortBy == null) ?
```
Comment (tinagrasso) [inline src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java]: Same comment as above, use Optionals.

**Review by tinagrasso — CHANGES_REQUESTED** [review]

File: src/test/java/com/google/step/finscholar/firebase/FirebaseStorageManagerTest.java [hunk 9]
```diff
@@ -41,18 +47,28 @@ public void isRecentlySetDocumentRetrievable() throws Exception {
...
-    Assert.assertNotEquals(ServletConstantValues.DEFAULT_VALUE, json);
+  public void setMultipleDocumentsAndRetrieveCollection() throws Exception {
+    FirebaseStorageManager.storeMultipleDocuments(firebase, TEST_COLLECTION_NAME, testObjectList);
+    String json = FirebaseStorageManager.getCollection(firebase, TEST_COLLECTION_NAME);
+    Optional<String> jsonNullable = Optional.ofNullable(json);
+    Assert.assertTrue(jsonNullable.isPresent());
+  }
+
+  @Test
+  public void retrieveCollectionBatchWithIDandWithoutID() throws Exception {
```
Comment (tinagrasso) [inline src/test/java/com/google/step/finscholar/firebase/FirebaseStorageManagerTest.java]: Break this into 2 tests.

```
retrieveCollecctionBatch_withId();

retrieveCollecctionBatch_withoutId();
```

File: src/test/java/com/google/step/finscholar/firebase/FirebaseStorageManagerTest.java [hunk 10]
```diff
@@ -15,9 +17,13 @@
 @RunWith(JUnit4.class)
 public class FirebaseStorageManagerTest {
   // Constants for FirebaseStorageManagerTest.
+  /** Logger that sends logs to the Cloud Project console. */
+  private static final Logger log = Logger.getLogger(FirebaseStorageManager.class.getName());
   public static final String TEST_COLLECTION_NAME = "testObjects";
   public static final String TEST_DOCUMENT_NAME = "testDocument";
   public static final String EXPECTED_DOCUMENT_RETRIEVABLE = "{\"one\":\"testObjects\",\"two\":\"testDocument\"}";
+  public static final int TEST_BATCH_SIZE_LIMIT = 10;
```
Comment (tinagrasso) [inline src/test/java/com/google/step/finscholar/firebase/FirebaseStorageManagerTest.java:40]: This batch size is larger than the number of test documents you have. Try to think of all the cases you need to  test to make sure you're grabbing the expected documents. I also don't see a test with no sort param set.

- first batch of collection where collection size > batch size
- middle batch of collection where collection size > batch size
- last batch of collection where # documents returns in < batch size

**Review by tinagrasso — CHANGES_REQUESTED** [review]
LG overall, but needs additional testing.

**Comment by MikolJiron** [comment]
So I added more test to make sure the batching code functions as intended! The one with successive batches in a row was especially important to make sure it worked.

File: src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java [hunk 11]
```diff
@@ -166,4 +168,192 @@ public static String getCollection(Firestore database, String collectionToGetFro
...
+   * Helper method for getCollectionBatch().
+   * Retrieve the first batch in the collection collectionReference of size batchSizeLimit.
+   * @param collectionReference - A reference to the collection I want to retrieve from.
+   * @param batchSizeLimit - The maximum size of the batch, which is the size of a new page of data in infinite scroll.
+   * @param parameterToSortBy - The parameter to sort the list of documents by.
+   * @return - The json string representing the first batch in a request for all of the documents in a collection.
+   * @throws FirebaseException
+   */
+  private static String getFirstBatch(CollectionReference collectionReference, int batchSizeLimit, String parameterToSortBy) 
+      throws FirebaseException {
```
Comment (Lorraine-Lyu) [inline src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java:218]: So should I directly pass in an optional object for id, batchSize and parameterToSortBy? We need to talk about this later today.

File: src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java [hunk 12]
```diff
@@ -111,7 +113,7 @@ public static void storeMultipleDocuments(Firestore database, String collectionT
    * @return - The JSON string representing the datapoint I just retrieved.
    */
   public static String getDocument(Firestore database, String collectionToGetFrom, String documentID)
```
Comment (Lorraine-Lyu) [inline src/main/java/com/google/step/finscholar/firebase/FirebaseStorageManager.java:115]: This id could be null, or an empty optional.

**Review by tinagrasso — APPROVED** [review]

**Pull request merged**